snuba/api.py

```diff
@@ -162,16 +162,80 @@ def query(validated_body=None, timer=None):
     )
 
 
+def format_query(dataset, body, table, prewhere_conditions, final) -> str:
```

Comment:
My goal is to eventually replace the `body, table, prewhere_conditions, final` part of this signature with a `Query` construct that is separate from the request body format. This will require (probably substantial) changes to `column_expr` and `condition_expr`.

Reply:
any chance we could have type hints in the signature (together with the result). It would make it easier to understand the method body.

Reply:
This seems so much more readable. One question at your plan for the Query construct. Are you going to create a class that takes body, dataset, table, pre_where and that has a method that returns the query string?
Second, how will this play with multiple dataset? Is every dataset going to provide its Query construct to allow for dataset specific logic in query parsing? Assuming we will allow datasets to provide their own logic.

Reply:
> any chance we could have type hints in the signature (together with the result). It would make it easier to understand the method body.

I can add them for `dataset`, `table`, and `final`, but the ones for `body` are going to take more work. I experimented a bit with this last week — anything that ends up using `column_expr` or `condition_expr` is going to take a lot of work to nail down. We can start partially typing `Query` into a `TypedDict` if you think that's helpful. I think we can at least have a comprehensive set of valid keys, and types that are more specific than `Any` for a handful of them (but unfortunately not really any of those that would be useful.)

Reply:
yeah, even Any would be reasonable for body, so that it sets expectations.